### ISSUE TITLE
unable to publish even in the --dry-run mode due to error

### ISSUE THREAD

**Issue by softstream-link** [rust-lang/cargo#12149]
### Problem

I have product called `byteserde` which is to `byte stream` what a `serde` is to `json`.

It contains two artifacts `byteserde` & `byteserde_derive` . The `byteserde_derive` artifact is meant to be used in the `byteserde` on certain internally declared types as well as in other external crates/projects. However in order for this to work I had to add this logic in my `byteserde_derive` artifact . It allows the `derive` to inject either a `crate::....` in front of `byteserde` dependencies when used in the ./src or a `::byteserde::.....` when used in the ./tests. It all works great until I try to package or publish the package.

```
pub fn get_crate_name() -> TokenStream {
    let cargo_crate_name = std::env::var("CARGO_CRATE_NAME").unwrap();
    let crate_name = match cargo_crate_name.as_str() {
        "byteserde" => quote!(crate),
        _ => quote!(::byteserde),
    };
    crate_name
}
``` 

### Steps

1. git clone git@github.com:softstream-link/byteserde.git
2. cd byteserde
3. cargo build && cargo test --examples --tests  # PASS
4. cargo package                                                      # FAIL
```
error[E0433]: failed to resolve: use of undeclared crate or module `byteserde`
  --> src/utils/strings/ascii/mod.rs:39:10
   |
39 | #[derive(ByteSerializeStack, ByteSerializeHeap, ByteDeserialize, PartialEq)]
   |          ^^^^^^^^^^^^^^^^^^ use of undeclared crate or module `byteserde`
   |
   = note: this error originates in the derive macro `ByteSerializeStack` (in Nightly builds, run with -Z macro-backtrace for more info)
```

### Possible Solution(s)

How can I enhance my `get_crate_name` logic to generate `crate::...` vs `::byteserde::` respectively and be able to package / publish?

### Version

latest 1.69

**Comment by weihanglo** [comment]
You may want to add `byteserde` as a dependency of `byteserde_derive`, similar to how `serde_derive` works. You can check their implementations as a reference.

As it doesn't look like a Cargo bug, I'd recommend asking this kind of question on [Rust user forum](https://users.rust-lang.org/) or elsewhere. Sorry we are unable to provide such support. This will be closed soon if you have no further questions. Thank you!

**Comment by softstream-link** [comment]
Please allow me to qualify what I think is somewhat unique problem.

In Classic Serde the crates dependency is as follows:
`serde`--optional-->`serde_derive` - this is because I shall be able to use code and traits in the `serde` crate manually and/or optionally rely on `serde_derive` to auto generate the code for typical/supported cases ( which of course is the preferred way )

In ByteSerde the crates dependency is different:

`byteserde` --required--> `byteserde_derive` - this is because the `byteserde` crate not only provides the serialization traits which can be used to manually write the code but also a few  `struct` types which are very typical when dealing with byte streams, ex: ascii string of fixed length. Now special/common type in the `byteserde` package relies on the `bytesede_derive` to auto generate its own serializers which creates the problem. 

1) when `byteserde_derive` crate is used by any third party package or end user on their own crate any code it spits outs need to be hygienic and hence any type references from the `byteserde` crate are created with absolute path so that it does not depend on user includes/use keywords.   Ex: `impl ::byteserde::ser::ByteSerialize for ArbitraryUserStruct { ... }`

2) on the other hand when `byteserde_derive` crate is used by `byteserde` and NOT a third party package the code it spits out can't use absolute path and must use relative path for `byteserde` to compile. Ex: `impl crate::ser::ByteSerialize for AsciiString {...}` where AsciiString is already part of the `byteserde` crate. 

3) when I write a `test` in the `byteserde` ./tests folder  the `byteserde_derive` code generation should revert back to treating the test target as a third party package since tests are compiled as entirely different crates. So the macro reverts back to generating absolute and not relative path, Ex: `impl::byteserde::ser::ByteSerialize for TestStructure {...}`

So in order to change the path I have added a check in the `byteserde_derive` crate on the env variable which works for builds. However, when I package or publish this does not seem to work since it seems to fallback to generating absolute path instead of relative path for case 2 above. 

So the question is whether there is a way for me to detect when cargo is running a package or publish command so that I can instruct the `byteserde_derive` crate to use relative path?

**Comment by epage** [comment]
> byteserde --required--> byteserde_derive - this is because the byteserde crate not only provides the serialization traits which can be used to manually write the code but also a few struct types which are very typical when dealing with byte streams, ex: ascii string of fixed length. Now special/common type in the byteserde package relies on the bytesede_derive to auto generate its own serializers which creates the problem.

Sounds like a workaround fro this would be to hand implement those traits.  This would also offerr the benefit of making the derive crate optional.

Personally, I'm getting the feeling that we do derive crates wrong.  If just one crate depends on `serde = { version = "1", features = ["derive"] }`, then all crates using `serde` can't build until all of the proc macro machinery is built, including `serde_json` or `toml` which don't use the derive.  In clap, I recently took the approach of making `clap` a facade that re-exports `clap_builder` and `clap_derive` though there are other ways of structuring this to solve the problem.

> So the question is whether there is a way for me to detect when cargo is running a package or publish command so that I can instruct the byteserde_derive crate to use relative path?

This question is exploring one way of solving the problem.  How do other proc macros that run into this problem solve it?

**Comment by softstream-link** [comment]
@epage - thanks for the suggestion, but I have three traits for ~5 types, can grow, in the `byteserde` crate which get autogenerated impl using `byteserde_derive`. In fact I did start to implement them manually in the beginning but it got old very quickly so I found a work around on stackoverflow which which path to use given value in the  `std::env::var("CARGO_CRATE_NAME")`. It seemed to work until I realized I can't publish. 

What is very interesting also is I have setup a local registry [cargo-http-registry](https://crates.io/crates/cargo-http-registry) so that I can try publishing `byteserde_derive` to it to see if I can make it work and to my surprise it just worked exactly like expected and not failed, which is what made me believe initially that this is a cargo for crates.io issue but not really sure.

```
cd byteserde_derive && cargo publish --registry local --allow-dirty
cd byteserde && cargo publish --registry local --allow-dirty
```

Another workaround that I am thinking about is to have yet a third crate, say `byteserde_types` where I only keep the types but then it would need to depend on both `byteserde` & `byteserde_derive` and if anyone wants to use it they would ultimately ended having to implicitly depend on all three anyway so breaking a `required` the dependency between `byteserde`  and `byteserde_derive` does not achieve anything.

Indeed not sure what other are doing to overcome this but would be very great full to get some pointers.
